### DISCUSSION TITLE
Meet pointwx, a beautiful Dash application to show off open-meteo data

### DISCUSSION THREAD

**Discussion by guidocioni** [open-meteo/open-meteo#548] [Show and tell]
I was waiting to have a pretty good working version of my application before sharing it but I'm not sure I'll ever reach a point where I'm fully satisfied so...here we go :) 

This is a fully-fledged python Dash (with bootstrap layout) multi-page application which uses `pandas` and `requests` to process data from the open-meteo API and turn them into beautiful plot using `plotly`. It is fully interactive, which means you can choose whatever combination of location and model you are interested in, but also zoom and pan on the plots to explore some specific features. It has built-in caching to avoid calling the open-meteo API to often and remembers the choice you made for location and models selection.
It shows not only high resolution forecasts for a location but also more interesting climate metrics which are usually hard to compute. 

Here are some screenshots.

**Ensemble forecasts**

![Screenshot 2023-12-12 at 09 19 58](https://github.com/open-meteo/open-meteo/assets/12760310/57d166f4-48bc-4d2d-9630-e127031f6147)

**Climate comparison for accumulated precipitation and temperature**

![Screenshot 2023-12-12 at 09 21 33](https://github.com/open-meteo/open-meteo/assets/12760310/69dd180b-ac2e-4f3c-9f08-dae466240323)

**Monthly Climatology**

![Screenshot 2023-12-12 at 09 21 20](https://github.com/open-meteo/open-meteo/assets/12760310/008a42df-97de-4b54-88d9-1cf5326353ea)

**Daily meteogram (based on Ensemble)**

![Screenshot 2023-12-12 at 09 20 44](https://github.com/open-meteo/open-meteo/assets/12760310/d8e675c7-8eb0-4b1f-b447-e5a46aa39e7b)

**Deterministic Forecasts (Multi-model Ensemble)**

![Screenshot 2023-12-12 at 09 20 30](https://github.com/open-meteo/open-meteo/assets/12760310/d5cd8aa0-4678-4fda-a39b-26daa1494cf9)

**Ensemble Heatmap**

![Screenshot 2023-12-12 at 09 20 15](https://github.com/open-meteo/open-meteo/assets/12760310/fb9e1a8a-3371-4bb7-8fad-8c55627ee64a)

Happy to try it? You can find the app here https://github.com/guidocioni/point_wx.
Please note this is an alpha version so there may be bugs. I'm happy to receive suggestions, forks, PRs... :) 
I'm still working on implementing new pages with new way of showing off the data, you can check the active issues.

Happy coding

Thanks to the support of Patrick you can try a live demo here https://hh.guidocioni.it/pointwx/. Given that it is deployed on premise and local resources it may not always stay online so use it at your own risk.

**Comment by guidocioni** [comment]
I've added a new page where you can create a pressure-time meteogram with the evolution of temperature, winds and cloud cover. Next stop, vertical soundings :) 

![newplot](https://github.com/open-meteo/open-meteo/assets/12760310/cf29f24d-a094-4ada-939a-057ddf099279)

**Reply by patrick-zippenfenig** [reply]
Really nice. I guess you have to list all pressure levels in the a very long URL. Would a synthesized variable like `temperature_all_hPa` or a range `temperature_200_to_1000_hPa` help?

**Reply by guidocioni** [reply]
Yes I do, but since I created an helper function to generate the URL (https://github.com/guidocioni/point_wx/blob/66d70cd761f976966a61fc8d9a1452a58802b14c/src/utils/openmeteo_api.py#L145) this is really a no-issue, as long as the URL does not exceed any critical length. 

Having a shortcut to get vertical data would be helpful but since it would only shorten the URL, but not change anything in terms of data transmitted and/or optimization on the server side, I would mark it as optional for now :)

**Reply by wreed1989** [reply]
Do you happen to know which packages you used to create this? Specifically for the following lines of import:

from .settings import cache, ENSEMBLE_VARS, ENSEMBLE_MODELS, DETERMINISTIC_VARS, DETERMINISTIC_MODELS, ROOT_DIR
from .custom_logger import logging, time_this_func

I was trying to replicate the code and create a similar plot with my own parameters but I keep running into a package error and can't track down which packages were installed prior to importing those modules. 

Thanks! Superbly clean plot!

**Reply by guidocioni** [reply]
did you install the packages listed in the README? sorry, the application is not properly packaged until now, but if you're having troubles running it you can open an issue in the repo and provide details on the error.
But anyway to make a plot you don't need all of that stuff, which is only required to run the application. You can isolate the functions defined in the `figures` module which always take care of drawing the plots (they all use `plotly`).

**Comment by Degamisu** [comment]
beautiful!

**Comment by guidocioni** [comment]
Just a heads up, you can now use a map to see the location selected.
I also implemented two additional ways of selecting a location:

- by clicking on the map
- by using geolocation 

<img width="657" alt="Screenshot 2024-02-02 at 11 05 28" src="https://github.com/open-meteo/open-meteo/assets/12760310/e05f559f-2bef-49e9-934a-6f52970f7d46">

**Comment by guidocioni** [comment]
I've added a new page with a simplified meteogram, where you can see the temperature and weather forecast for the next days. I'm using ensemble data because this way I can actually compute the spread of the forecast (which, as expected, increases with time).
Depending on the model you can also see sunshine hours and precipitation amount (together with range and probability).

![newplot](https://github.com/open-meteo/open-meteo/assets/12760310/c4f826f2-519c-4159-b155-d59e3055751e)

The cool thing is that you can directly compare these values with the diamonds (climatology) to assess whether the forecast is anomalous or not. 

I still want to implement
- Forecast uncertainty as a value
- Wind speed and direction daily ranges
- Total precipitation with snow/rain parts
- Better way to compute the daily weather code. For the moment I'm taking the median, because it seems to be more relevant than the mean and the max, but I still get too many "cloudy" days. 

You can directly test this feature here https://hh.guidocioni.it/pointwx/meteogram

**Reply by patrick-zippenfenig** [reply]
That looks really good! For forecast uncertainty the new (not yet officially announced) [previous runs API](https://open-meteo.com/en/docs/previous-runs-api/) could be a good start to analyse the "stability" of the forecast. With this data you could also show the relative forecast skill for each weather model independently

**Reply by guidocioni** [reply]
I don't have time to polish a new feature that you already publish a new endpoint 😄 amazing!
For the uncertainty I was thinking more about combining the std of temperature, winds, pressure....into a single metric (similar to what meteoblue does for their "predictability" index).

**Comment by winash12** [comment]
Out of curiosity are these interactive i.e. do they use callbacks ? I watnt to move a cursor over a curve and have it tell me the temperature or dew point value etc.

**Reply by guidocioni** [reply]
you don't have to write a callback to do that, you can specify the value of the tooltip pretty easily 😉

**Comment by guidocioni** [comment]
I've implemented a preliminary SkewT diagram that allows one to evaluate the atmospheric profile and animate it. 

In the video an example of the inversion layer in Hamburg being eroded in the next few days (finally!) by the arrival of upper-level cold air.

https://github.com/user-attachments/assets/c22168d1-3c1b-4ba7-8bf8-01443b7455d3

Available as usual at https://hh.guidocioni.it/pointwx/vertical

**Comment by winash12** [comment]
Isn't a sounding plot available as part of MetPy ?

**Reply by guidocioni** [reply]
Yes but this is interactive (and made in plotly).
To my knowledge there's no package that makes interactive skewt plot in Python (besides Sharppy).
I still used some functions from metpy to compute the parcel profiles.

**Comment by winash12** [comment]
Yes thanks. I need to download your code and install it on my Linux desktop. From that movie you have shared I can't figure out what is interactive about that sounding plot. Are you saying you can change the date and location on the fly ?

**Reply by guidocioni** [reply]
The interactive part consists in zooming, panning and in general getting information on the data that you won't be able to get from a simple matplotlib plot. You can test the application online, there's no need to deploy your own copy for testing.

**Comment by winash12** [comment]
That sounds fantastic. I am wanting to use the https://weather.uwyo.edu/upperair/sounding.html soundings. Is that possible with your code ? I guess I need to use a combination of siphon , metpy and your code to deliver a sounding plot.

**Reply by guidocioni** [reply]
If the goal is just to interact with the data yourself then SharpPy is definitely the best option.
Otherwise yes I guess you could strip out the function that does the SkewT plot from my code and use it with Wyoming sounding data. You just need to pass the right data to the function to make that work, or just adapt the function. 
However this is out of the scope of my application ;)
